### PR TITLE
CI: systest: safer random_rfc1918_subnet

### PR DESCRIPTION
**Pull request by edsantiago**
Our aarch64 CI system uses 172.31.0.0/20. Because I was (and am)
lazy, my random_rfc1918_subnet() helper was only checking /24.
This causes flakes.

Solution is to actually do it right: binary arithmetic, prefix
matching. This is effectively impossible in bash, so, use a
hairy perl oneliner and add copious tests.

Note: I considered `ipcalc`, `ipcalc-ng`, `sipcalc`, various python packages. Given our system-test constraints where we need to run on fedora, debian, and RHEL, I believe this is the safest approach in terms of portability. There may be cleaner options I did not consider, so please speak up if you have better suggestions.

Fixes: #18693

Signed-off-by: Ed Santiago <santiago@redhat.com>
```release-note
None
```

### CODE REVIEW

File: test/system/helpers.network.bash [hunk 1]
```diff
@@ -140,6 +142,32 @@ function random_rfc1918_subnet() {
...
+    #    127.0.0.0/1   -> 0
+    #    128.0.0.0/1   -> 1
+    #    10.0.0.0/8    -> 00001010
+    #
+    # I'm so sorry for the ugliness.
+    while read prefix; do
+        if [[ "$binaddr" =~ ^$prefix ]]; then
+            return 0
+        fi
+    done < <(ip route list | perl -ne 'if (m!^([\d.]+)/(\d+)!) { @ip = split(/\./, $1); $bits = $2; $bits = 24 if $bits > 24; $mask = (2 ** $bits - 1) << (32 - $bits); $ip = unpack("N",pack("C4",@ip)); $net = sprintf("%032b", ($ip & $mask)); printf "%.*s\n", $bits, $net }')
```
Comment (Luap99) [inline test/system/helpers.network.bash:166]: it is extremely hard to follow that line, can you somehow pretty print it instead of smashing into one gigantic one liner?

Also in case it helps you parsing the output you could run `ip -j route` which gives you json so you wouldn't need the more complex regex to filter, i.e. `ip -j route | jq -r .[].dst` would look much simpler to me to filter the destination addresses, of course you still need to filter out "default" from that list.

Comment (edsantiago) [reply test/system/helpers.network.bash]: Rewritten and repushed. The rewrite let me see unnecessary code that I have now excised, yay.

Tests running simultaneously now here and in #17831 and #20161.

**Review by Luap99 — APPROVED** [review]
Thank you, it is much more readable and understandable now!
LGTM

File: test/system/helpers.network.bash [hunk 2]
```diff
@@ -140,6 +142,69 @@ function random_rfc1918_subnet() {
...
 }
 
+# subnet_in_use() - true if subnet already routed on host
+function subnet_in_use() {
+    local subnet_script=${PODMAN_TMPDIR-/var/tmp}/subnet-in-use
+    rm -f $subnet_script
+
+    # This would be a nightmare to do in bash. ipcalc, ipcalc-ng, sipcalc
+    # would be nice but are unavailable some environments (cough RHEL).
+    # Likewise python/perl netmask modules. So, use bare-bones perl.
```
Comment (TomSweeneyRedHat) [inline test/system/helpers.network.bash:152]: Perl! My first true scripting love!  :^)

**Comment by TomSweeneyRedHat** [comment]
LGTM
